src/main/scala/zio/s3/Live.scala

```diff
@@ -105,78 +108,78 @@ final class Live(unsafeClient: S3AsyncClient) extends S3.Service {
           )
         )
       )
       .unit
 
-  // only for file which are bigger than 5 Mb
   def multipartUpload[R <: zio.Has[_]: Tag](
     bucketName: String,
     key: String,
-    contentType: String,
     content: ZStream[R, Throwable, Byte],
-    metadata: Map[String, String] = Map.empty
+    options: UploadOptions = UploadOptions.default,
+    partSize: Int = UploadOptions.MinMultipartPartSize,
+    parallelism: Int = 1
   ): ZIO[R, S3Exception, Unit] =
     for {
-      uploadId <- execute(
-                    _.createMultipartUpload(
-                      CreateMultipartUploadRequest
-                        .builder()
-                        .bucket(bucketName)
-                        .key(key)
-                        .contentType(contentType)
-                        .metadata(metadata.asJava)
-                        .build()
-                    )
-                  ).map(_.uploadId())
-
-      parts    <- content
-                  //part size limit is 5Mb, required by amazon api
-                    .chunkN(5 * 1024 * 1024)
-                    .mapChunks(Chunk.single)
-                    .zipWithIndex
-                    .mapM {
-                      case (chunk, partNumber) =>
-                        execute(
-                          _.uploadPart(
-                            UploadPartRequest
-                              .builder()
-                              .bucket(bucketName)
-                              .key(key)
-                              .partNumber(partNumber.toInt + 1)
-                              .uploadId(uploadId)
-                              .contentLength(chunk.length.toLong)
-                              .build(),
-                            AsyncRequestBody.fromBytes(chunk.toArray)
-                          )
-                        ).map(r => CompletedPart.builder().partNumber(partNumber.toInt + 1).eTag(r.eTag()).build())
-                    }
-                    .runCollect
-                    .mapError(S3ExceptionUtils.fromThrowable)
-
-      _        <- execute(
-                    _.completeMultipartUpload(
-                      CompleteMultipartUploadRequest
-                        .builder()
-                        .bucket(bucketName)
-                        .key(key)
-                        .multipartUpload(CompletedMultipartUpload.builder().parts(parts.asJavaCollection).build())
-                        .uploadId(uploadId)
-                        .build()
-                    )
-                  )
+      uploadId      <- execute(
+                         _.createMultipartUpload(
+                           UploadOptions
+                             .multipartUploadBuilder(options)
+                             .bucket(bucketName)
+                             .key(key)
+                             .build()
+                         )
+                       ).map(_.uploadId())
+
+      parts         <- content
```

Comment:
that should handle the empty case, need to test with amazon api if you can send send request  with same part number 

(ZStream.succeed(Chunk.empty -> 0L) ++ content
                  //part size limit is 5Mb, required by amazon api
                    .chunkN(5 * 1024 * 1024)
                    .mapChunks(Chunk.single)
                    .zipWithIndex)

Reply:
the problem with this approach is the following: if the stream is actually not empty, it won't work.
For example, we have a real chunk of size 1MB. And we prepend 0 bytes chunk. So now we have 2 chunks: [0, 1Mb].
The last chunk can be smaller than 5MB. But while we have 2 chunks, so the call will send an exception, because the first chunk is also smaller than 5 MB.

Akka streams has a nice operator for this https://doc.akka.io/docs/akka/current/stream/operators/Source-or-Flow/orElse.html. But I haven't found a similar operator in zio.

Reply:
- if stream is empty we are sending an empty byte (same as your solution)  => s3 sending OK
- if stream is not empty inferior of 5Mb, we are sending 0bytes, and after 1Mb => s3 sending FAIL when 1Mb is received
- if stream is 6Mb, we are sending 0bytes, and after 6Mb => s3 sending OK

are we agree ?

Reply:
> if stream is not empty inferior of 5Mb, we are sending 0bytes with PartNumber 1, and after 1Mb with PartNumber 1 => s3 sending FAIL when 1Mb is received
yes

> if stream is 6Mb, we are sending 0bytes with PartNumber 1, and after 6Mb with PartNumber 1 => s3 sending OK
the problem here is that 6MB is actually split into 2 chunks: 1MB and 5MB

Reply:
sorry i wanted to say 

if stream is 6Mb, we are sending 0bytes with PartNumber 1, 5Mb with PartNumber 1 and 1Mb with partNumber 2 => s3 sending OK

Reply:
here is an implementation without an explicit call https://github.com/zio/zio-s3/pull/117/commits/d6b2f1b0bb06dfc48d245b6409ab4654c6479d82